Port stereo_image_proc to ROS 2

jacobperron opened the pull request:
Connects to #360. 

I think the majority of the port is complete. Opening this PR for visibility/early feedback. I've removed OpenCV 2 support in an attempt to simplify the code.

In addition to the port, I've added some simple launch tests that can be expanded upon in the future.
Test data taken from OpenCV ([from here](https://github.com/opencv/opencv_extra/tree/dc0c6c1ee2cd142f936e7f957bbc595df2ce17e8/testdata/gpu/stereobm)).

Here's what's missing:

- Monitoring the status of connected publishers to dynamically subscribe/unsubscribe as needed

https://github.com/ros-perception/image_pipeline/blob/a0a361be0eb6cfac94d18ebc53f4a4f906dfe0e1/stereo_image_proc/src/stereo_image_proc/disparity_node.cpp#L128-L132

This feature doesn't exist in ROS 2, so I've left TODOs in the code (similar to what was done in other image_pipeline packages).

- Reconfigurable callbacks for parameters. This should be achievable since parameters are configurable by default. I'll proceed replacing the use of dynamic reconfigure using the rclcpp API. I can add it to this PR or after it is merged, whatever comes first.

- I should also double-check QoS compatibility with other parts of image_pipeline (mentioned in https://github.com/ros-perception/image_pipeline/issues/360#issuecomment-564571781).

- Windows support. E.g add missing visibility flags, cmake commands, etc.

---

I've been building/testing with ROS 2 Eloquent and the latest ROS 2 source code.

jacobperron committed:
Rename source files

More descriptive names that will match the classes contained inside.
Also moved all source files associated with the library into a directory of the same name.

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Get stereo_image_proc compiling with ROS 2

Made some minimum changes required for compilation. Not feature-complete yet.

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Add launch test for disparity node

The test simply checks if the disparity node publishes to the expected topic when it received input.
Made some fixes added to CMakeLists.txt to get node to run probably and pass the test.

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Add launch test for PointCloudNode

Similar to the test for the disparity node.
Fix bug in PointCloudNode.

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Remove unused files

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Add Python launch file

Replaces the old ROS 1 XML launch file.

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Remove standalone executable

This form of manual composition seems redundant to me and can be done with a launch file.
The only thing it appeared to add was the 'advertisement checker', which was not ported.

If there is strong objection to removing this executable, we can add it back.

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Add linter tests and lint

There are a few cases where we discard the const-qualifier for the image data.
This happens because OpenCV won't accend const pointers to data.
Rather than changing the API, I've introduced a const_cast instead.

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Remove OpenCV 2 support

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Expand macros

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Enable uncrustify test

Fix lint error.

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Remove dead code and add TODO

The code removed was never being used.
Added a TODO regarding the use of an unset variable. This has been here since the ROS 1 implementation,
I'm not sure what the logic is intended to do so I've left it for now.

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Update package.xml version

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Minor refactor

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Fix typo

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Minor: update include syntax and order for consistency

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Fix rosdep key typo

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron committed:
Move inline implementation into class

Save on some vertical space.

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron commented:
`rclcpp_components_register_node` was [recently backported to Dashing](https://github.com/ros2/rclcpp/pull/935). It should be available in the [next patch release](https://discourse.ros.org/t/preparing-for-dashing-sync-and-patch-release-2019-12-12/11791) and then hopefully CI will be happy.

SteveMacenski commented on stereo_image_proc/launch/stereo_image_proc.launch.py:
```diff
@@ -0,0 +1,64 @@
+
+
+def generate_launch_description():
+    # TODO(jacobperron): Include image_proc launch file when it exists
```
Something blocking this?

jacobperron replied:
Not that I'm aware of. I think we should be able to port [this launch file](https://github.com/ros-perception/image_pipeline/blob/568bb54404b740295a505234b08e23db5ddc22a3/image_proc/launch/image_proc.launchhttps://github.com/ros-perception/image_pipeline/blob/568bb54404b740295a505234b08e23db5ddc22a3/image_proc/launch/image_proc.launch), I just didn't want to add it to this PR.

SteveMacenski replied:
OK, just wondering

SteveMacenski commented on stereo_image_proc/src/stereo_image_proc/disparity_node.cpp:
```diff
@@ -0,0 +1,240 @@
+  int left = block_matcher_.getDisparityRange() + block_matcher_.getMinDisparity() + border - 1;
+  int wtf;
+  if (block_matcher_.getMinDisparity() >= 0) {
+    wtf = border + block_matcher_.getMinDisparity();
```
What does `wtf` represent?  I think that could be named clearer

jacobperron replied:
I wish I knew, haha. I didn't touch the logic that was here since ROS 1. Any suggestions are appreciated.

SteveMacenski replied:
I... don't even know what that is doing (or why there isn't one of those for the height). I'm going to just back away slowly before I go too far down a rabbithole

SteveMacenski commented on stereo_image_proc/src/stereo_image_proc/point_cloud_node.cpp:
```diff
@@ -0,0 +1,268 @@
+  std::shared_ptr<ApproximateSync> approximate_sync_;
+
+  // Publications
+  // TODO(jacobperron): Uncomment when we can be notified of subscriber status
```
Can't you check with http://docs.ros2.org/dashing/api/rclcpp/classrclcpp_1_1Node.html#a7aa2ed22450d2f3300ea5b879acd73b8?

jacobperron replied:
We can poll for subscriptions, but I think we'd prefer this to be event driven.
I wasn't aware of the graph event API, I'll see if we can use it.

SteveMacenski replied:
Ah, well what I think would accomplish the exact same thing would be to have the callback just check if any of the output topics have any subscribers, else just return without doing any work.

The `connectCb()` is nice so you don't have to process any of the callbacks that would just be exiting, but right now, Dashing/Eloquent have an issue with unsubscribing and resubscribing to a topic causing a crash (we have some navigation / rclcpp tickets around), so I'd rather do it in a non-crashing way. And its a way we can do right now with that API call. Some small performance hit, but the period of time a pipeline is A) up B) gets a message and C) doesn't have any users is pretty short

jacobperron replied:
Sure, that sounds reasonable. For posterity, could you reference an issue regarding problems with resubscribing? Thanks!

SteveMacenski replied:
https://github.com/ros-planning/navigation2/issues/1363

https://github.com/ros2/rmw_fastrtps/issues/341

jacobperron replied:
See 3e40a3d

SteveMacenski reviewed (changes requested):
LGTM, just a couple comments

jacobperron committed:
Avoid extra computation if there are no matching subscriptions

Remove commented code related to subscriber status since this may look different in ROS 2.

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron commented:
Note, I also did some minor refactoring a25372d

SteveMacenski reviewed (approved):
Any other changes you'd like to make?

We should wait until we pass CI though,

jacobperron commented:
I want to add the reconfigurable parameters, and figure out what `wtf` is.. but I can make changes in separate PRs.

It'd be nice if CI was green though before this one goes in. I figured since the latest dashing sync it should be working... :thinking:

JWhitleyWork commented:
@jacobperron / @SteveMacenski - They have to manually re-build the Docker containers with the updated packages for CI to pass.

SteveMacenski commented:
Looks like it was updated 3 days ago https://hub.docker.com/_/ros?tab=tags

SteveMacenski commented:
Potentially do it the more nasty way to get CI to build?

JWhitleyWork commented:
@SteveMacenski - You're right. I checked the SHA hash of the image it pulled and CI must be using a cached copy somewhere. I don't know that there is anyway to update this from our end.

jacobperron commented:
I don't think the image has been updated yet. If I pull `ros:dashing-ros-base` locally, it doesn't contain the latest version of [ros_base](https://github.com/ros2/variants/tree/dashing/ros_base) (0.7.3).

SteveMacenski commented:
https://circleci.com/docs/2.0/caching/#clearing-cache

@jacobperron  try adding a cache variable we can increment, ei `CACHE_VERSION=v1`

Edit: ah, ok

jacobperron commented:
Also, I just noticed that my tests aren't compatible with Dashing, so I'll update those in the meantime.

jacobperron committed:
Refactor tests for compatbility with Dashing

Now the launch tests should work for both Dashing and Eloquent.
I left some TODO's where we can simplify once we diverge from Dashing.

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron commented:
@SteveMacenski I've updated the tests so they're backwards compatible with Dashing (ecce41b). PTAL.

jacobperron committed:
Use default context

Signed-off-by: Jacob Perron <jacob@openrobotics.org>

jacobperron commented:
TL;DR It doesn't seem like the Docker image will be updated any time soon. Not until the underlying Ubuntu image gets rebuilt: https://github.com/osrf/docker_images/issues/112

It's a bit annoying, but I can refactor the CMakeLists.txt to see if we can make CI happy.

jacobperron commented:
See https://github.com/ros-perception/image_pipeline/pull/487 for a proposed fix for failing CI.

SteveMacenski commented:
![](https://media1.tenor.com/images/9fa86c06636a5d1c451298f1c339bcc5/tenor.gif?itemid=12807779)

Merged.

jacobperron commented:
Rebased :crossed_fingers:
